!#/bin/bash
# useful bedtools commands
# bedtools v2.24.0

# get a count of overlaps in filtered data set
# http://bedtools.readthedocs.io/en/latest/content/tools/intersect.html
#We start by running BEDtools on the alignment set restricted to the specified region (Chr3:11777000-11794000) and the GTF annotation file listed above. To allow the input to be read directly from the BAM file, we use the option ‘-abam’; in this case we will need to also specify ‘-bed’ for the BAM alignment information to be shown in BED column format in the output:



